### ISSUE TITLE
CLP(FD) request: improve propagation for exponentiation

### ISSUE THREAD

**Issue by JCumin** [SWI-Prolog/swipl-devel#167]
I have identified a situation where I believe CLP(FD) could reduce the search space much more than what it currently does:

```
?-  X^Y #= 0.
X^Y#=0.
```

I believe there is only one possible value for `X` for which `X^Y = 0`: `0`. Therefore `X^Y #= 0` could directly unify `X` with `0`.

Similarly, for:

```
?- X in 0..100, 2^I #= X.
X in 0..100,
2^I#=X.
```

The domain of `X` could be reduced to `1..100`.

This leads to really long computing times for seemingly trivial queries, such as:

```

?- time((I in 0..10000, 2^I #= 0, indomain(I))).
% 1,330,353 inferences, 1.576 CPU in 1.577 seconds (100% CPU, 844342 Lips)
false.
```

**Comment by jburse** [comment]
0^1 = 0.

**Comment by triska** [comment]
A commit I recently installed seems to have improved the test case a lot, please try it out:

https://github.com/SWI-Prolog/swipl-devel/commit/7a77b2984ec58c3c3261ce091870e278c0c09987

The propagation can still be improved!

**Comment by JCumin** [comment]
@triska i only see a difference for the following case (and none for the other two cases with `… #= 0`):
```

?- X in 0..100, 2^I #= X.
X in 0..100,
2^I#=X,
I in inf..6
```

However I don't understand why `I`'s lower bound is `inf` instead of `0`. The only possible cases, as far as I know, where `I in inf.. -1, X^I = Y` with `X` and `Y` integers is when `X = Y = 0` or `X = Y = 1`, which can both be ruled out because `X = 2` here.

**Comment by triska** [comment]
Yes, you are right: These cases can of course be ruled out. It is "only" a matter of adding these rules to the library.

However, doing so is extraordinarily hard and error-prone, so not everything that *can* be propagated *is* propagated because not all cases are implemented. Especially for highly non-linear constraints, it takes a lot of effort to think all cases through and remove only those domain elements that cannot participate in solutions.

You are one of the few users of CLP(FD) who make improvements possible, because you state simple cases in which you expect better propagation. Such cooperation can almost be called old-school at this point, although it has been for many years standard practice for&nbsp;CLP(FD). For this reason, I am still making some improvements on the SWI version of CLP(FD) in those cases you need. If you want to try your luck with adding additional propagation rules that you detect (in this case: for `pexp`, around l.&nbsp;5000 of `clpfd.pl`), please file a pull request.

*Le&nbsp;Graal de propagation*, if we can call it that, will consist in finding a way to *generate* these propagation rules automatically, ideally starting from a declarative definition of what we mean by "exponentiation" (in this case). Currently, they are only implemented manually, by painstakingly considering the different cases.

**Comment by JCumin** [comment]
@triska Thanks!

**Comment by triska** [comment]
I am reopening this because the issue is not solved!
